### PR TITLE
Add number of Instances in creating and deleting states to gce cache

### PR DESCRIPTION
**Pull request by damikag**
This PR adds the number of instances of a MIG in creating and deleting states to gce cache.

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds ```migInstancesCreatingCount``` and ```migInstancesDeletingCount``` to gce cache.
Number of instances of MIGs in creating and deleting states are cached when calling ```gce.fillMigInfoCache()```
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?
NONE

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/gce/mig_info_provider.go [hunk 1]
```diff
@@ -359,6 +359,8 @@ func (c *cachingMigInfoProvider) fillMigInfoCache() error {
 				c.cache.SetMigTargetSize(zoneMigRef, zoneMig.TargetSize)
 				c.cache.SetMigBasename(zoneMigRef, zoneMig.BaseInstanceName)
 				c.cache.SetListManagedInstancesResults(zoneMigRef, zoneMig.ListManagedInstancesResults)
+				c.cache.SetMigInstancesCreatingCount(zoneMigRef, zoneMig.CurrentActions.Creating)
```
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/gce/mig_info_provider.go:362]: Cluster Autoscaler has a different concept of creating and deleting than the gce instance. Please take a look at this code: https://github.com/kubernetes/autoscaler/blob/8040eadfe4f4287a14da7b3f703384c11c98dd86/cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go#L492

We need to also count instances in states other that creating and deleting.

**Review by bskiba — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/cloudprovider/gce/cache.go [hunk 2]
```diff
@@ -91,6 +93,8 @@ func NewGceCache() *GceCache {
 		machinesCache:                    map[MachineTypeKey]MachineType{},
 		migTargetSizeCache:               map[GceRef]int64{},
 		migBaseNameCache:                 map[GceRef]string{},
+		migInstancesCreatingCount:        map[GceRef]int64{},
```
[the remark this answers is not in the document]
Comment (damikag) [reply cluster-autoscaler/cloudprovider/gce/cache.go:96]: Updated

File: cluster-autoscaler/cloudprovider/gce/mig_info_provider.go (hunk 1 above)
Comment (damikag) [reply cluster-autoscaler/cloudprovider/gce/mig_info_provider.go:362]: Updated

File: cluster-autoscaler/cloudprovider/gce/mig_info_provider.go [hunk 3]
```diff
@@ -444,3 +446,25 @@ func (c *cachingMigInfoProvider) GetListManagedInstancesResults(migRef GceRef) (
...
 	return listManagedInstancesResults, nil
 }
+
+func createInstancesState(actionsSummary *gce.InstanceGroupManagerActionsSummary) map[cloudprovider.InstanceState]int64 {
+	state := map[cloudprovider.InstanceState]int64{
+		cloudprovider.InstanceCreating: 0,
+		cloudprovider.InstanceDeleting: 0,
+		cloudprovider.InstanceRunning:  0,
+	}
+	state[getInstanceState("ABANDONING")] += actionsSummary.Abandoning
```
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/gce/mig_info_provider.go:456]: The thing that worries me about this approach is that if a new action is added, we will not count those instances at all and have no idea.

So we need to either
a) make sure that no actions will ever be added to the list of possible vm actions
b) just get the counts for "interesting actions" ("CREATING", "RECREATING", "CREATING_WITHOUT_RETRIES", "ABANDONING", "DELETING") from actionsSummary and calculate the running instances as targetSize - creating. This requires confirmation that targetSize = #instances_in_all_states - #instances_deleting - #instances_abandoning

OTOH, do we need the exact number of running instances? Maybe we just want to keep the total instances count? That number will AFAICT be equal to targetSize + #instances_deleting + #instances_abandoning, although this needs to be confirmed as well

Comment (damikag) [reply cluster-autoscaler/cloudprovider/gce/mig_info_provider.go:456]: Implemented option b)

File: cluster-autoscaler/cloudprovider/gce/mig_info_provider.go [hunk 4]
```diff
@@ -444,3 +446,18 @@ func (c *cachingMigInfoProvider) GetListManagedInstancesResults(migRef GceRef) (
 	c.cache.SetListManagedInstancesResults(migRef, listManagedInstancesResults)
 	return listManagedInstancesResults, nil
 }
+
+func createInstancesState(targetSize int64, actionsSummary *gce.InstanceGroupManagerActionsSummary) map[cloudprovider.InstanceState]int64 {
```
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/gce/mig_info_provider.go:450]: We need a test for this method. I know there is currently no good public method to test this with, but in the meantime, please add a test for createInstanceState

**Review by bskiba — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/cloudprovider/gce/cache.go [hunk 5]
```diff
@@ -546,3 +548,25 @@ func (gc *GceCache) InvalidateAllListManagedInstancesResults() {
 	defer gc.cacheMutex.Unlock()
 	gc.listManagedInstancesResultsCache = make(map[GceRef]string)
 }
+
+// GetMigInstancesStateCache returns instancesState for the given mig from cache.
+func (gc *GceCache) GetMigInstancesStateCache(migRef GceRef) (instanceState map[cloudprovider.InstanceState]int64, found bool) {
```
[the remark this answers is not in the document]
Comment (damikag) [reply cluster-autoscaler/cloudprovider/gce/cache.go:553]: Done

File: cluster-autoscaler/cloudprovider/gce/mig_info_provider.go (hunk 4 above)
Comment (damikag) [reply cluster-autoscaler/cloudprovider/gce/mig_info_provider.go:450]: Done

File: cluster-autoscaler/cloudprovider/gce/mig_info_provider_test.go [hunk 6]
```diff
@@ -612,10 +612,12 @@ func TestRegenerateMigInstancesCache(t *testing.T) {
 
 func TestGetMigTargetSize(t *testing.T) {
 	targetSize := int64(42)
+	igmSummary := gce.InstanceGroupManagerActionsSummary{Creating: 0, Deleting: 0}
```
[the remark this answers is not in the document]
Comment (damikag) [reply cluster-autoscaler/cloudprovider/gce/mig_info_provider_test.go:615]: Done

**Pull request merged**